0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,463 @@
+
+```proto
+// This is intended for the Scheduler Service to call back to the daprd sidecar at the time of the job schedule... ie it's trigger time.
+service SchedulerCallback {
+
+    // Callback RPC for job schedule being at 'trigger' time
+    rpc TriggerJob(TriggerRequest) returns (google.protobuf.Empty) {}
+}
+
+// TriggerRequest is the message request to convey that the job is at its 'trigger' time. 
+message TriggerRequest {
+    // The job. 
+    Job job = 1; 
+
+    // The metadata associated with the job.
+    // This can contain the generated `key` for the optional state store when the daprd sidecar needs to lookup the entire data from a state store.
+    // The sidecar will use the metadata in order to know whether this job is registered for an actor. This is needed, as the routing mechanism for actors is different for the callback.
+    map<string,string> metadata = 2;
+}
+```
+
+![CallBack](./resources/0004-BIRS-distributed-scheduler/callback.png)
+
+### Scheduler Service
+
+![Pluggable Scheduler Service](./resources/0004-BIRS-distributed-scheduler/pluggableSchedulerService.png)
+
+A new `Scheduler Service` is created in the control plane. This Scheduler Service will include an embedded etcd instance (persisted) as well as a Scheduler Dapr API, which will live in `dapr/proto/scheduler/scheduler.proto`. The Scheduler Service is pluggable and allows for different implementations as needed. It is installed by default into the local development environment on `dapr init` similar to other control plane services in Dapr. This is an optional service and runs in a local container.
+
+To guarantee we don't have several Scheduler Service instances firing off the same job, we will have **virtual partitioning** (in-memory) such that each Scheduler Service instance owns a subset of all the jobs that exist. 
+
+### CLI
+
+`dapr schedule <name> –schedule “@hourly” –data “<cloud-event-json>”`
+
+### Implications
+
+- The Scheduler Building Block and Service will result in the ***deprecation*** of `actor reminders` and the `bindings.cron` component.
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+    * To start, this will implement generic scheduler logic, then will be expanded to enable the:
+        * Delayed PubSub
+        * Scheduled Service Invocation
+        * Actor Reminders
+* What alternatives have been considered, and why do they not solve the problem?
+    * Placement with Actor reminders is *very* Actor specific. The goal is to have a new scheduler in Dapr that is reusable by several building blocks.
+* What advantages does this proposal have? 
+    * This design has the advantage of enabling the Scheduler Service to be implemented in different ways. 
+    * This design also enables flexibility in where the data is stored. Whether that is in etcd in full, or partially with a reference to a state store (component) of the user's choice.
+
+### Acceptance Criteria
+
+* How will success be measured? 
```

Comment:
Please also add an acceptance criteria to test what happens in disaster scenarios (when more than n/2-1 instances fail, such as 2 out of 3). Also need to test how it recovers from total cluster shutdowns - I know of more than a few users that shut down their  development K8s clusters when not in use, for example.

Reply:
There can be multiple disaster scenarios that are hard to predict, so I will add a backup & recovery scenario to be documented to the AC.

Reply:
There can be multiple failure scenarios but the one that is most critical is failure of N/2+1 nodes or every node (so with 3 nodes, failure of 2 or 3). This can be tested easily by shutting down some of the nodes.